File: archive/f/f-sharp/Zeckendorf.fs
open System

module Zeckendorf =
    let fibsUpTo n =
        let rec build acc a b =
            if a > n then acc else build (a :: acc) b (a + b)

        build [] 1 2 |> List.toArray

    let run n =
        if n = 0 then
            ""
        else
            let fibs = fibsUpTo n

            Array.foldBack
                (fun f (remaining, acc) ->
                    if f <= remaining then
                        (remaining - f, f :: acc)
                    else
                        (remaining, acc))
                fibs
                (n, [])
            |> snd
            |> List.map string
            |> String.concat ", "

module Helpers =
    let usage = "Usage: please input a non-negative integer"

    let (|NonNegativeInt|_|) (s: string) =
        match s.Trim() |> Int32.TryParse with
        | true, n when n >= 0 -> Some n
        | _ -> None

    let parseArgs =
        function
        | [| NonNegativeInt n |] -> Ok n
        | _ -> Error usage

    let handleResult =
        function
        | Ok result ->
            printfn "%s" result
            0
        | Error msg ->
            eprintfn "%s" msg
            1

[<EntryPoint>]
let main argv =
    argv |> Helpers.parseArgs |> Result.map Zeckendorf.run |> Helpers.handleResult

File: archive/o/ocaml/zeckendorf.ml
let max_2_fib_up_to n =
  let rec aux prev curr =
    if curr + prev > n then (curr, prev) else aux curr (prev + curr)
  in
  aux 1 1

let zeckendorf n =
  let rec aux high_f low_f rem acc =
    if rem = 0 then List.rev acc
    else
      let new_rem, new_acc =
        if high_f <= rem then (rem - high_f, high_f :: acc) else (rem, acc)
      in
      aux low_f (high_f - low_f) new_rem new_acc
  in
  let high_f, low_f = max_2_fib_up_to n in
  aux high_f low_f n []

let num_list_str nums = nums |> List.map string_of_int |> String.concat ", "

let parse_args argv =
  match argv with [| _; n |] -> int_of_string_opt n | _ -> None

let () =
  print_endline @@
  match parse_args Sys.argv with
  | Some num when num >= 0 -> num_list_str (zeckendorf num)
  | _ -> "Usage: please input a non-negative integer"
